src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java

```diff
@@ -201,23 +213,31 @@ private Object traverseField(String fieldName, Value fieldValue) {
                 .collect(Collectors.toList());
     }
 
-    private void traverseField(IMessage message, String fieldName, Value value, IFieldStructure fieldStructure) {
-        if (fieldStructure != null) {
-            Object convertedValue = fieldStructure.isComplex()
-                    ? processComplex(value, fieldStructure)
-                    : convertSimple(value, fieldStructure);
-            message.addField(fieldName, convertedValue);
-        }
+    private void traverseField(IMessage message, String fieldName, Value value, @NotNull IFieldStructure fieldStructure) {
+        Object convertedValue = fieldStructure.isComplex()
+                ? processComplex(value, fieldStructure)
+                : convertSimple(value, fieldStructure);
+        message.addField(fieldName, convertedValue);
     }
 
-    private Object convertSimple(Value value, IFieldStructure fieldStructure) {
+    private Object convertSimple(Value listEvent, IFieldStructure fieldStructure) {
         if (fieldStructure.isCollection()) {
-            return value.getListValue().getValuesList()
-                    .stream()
-                    .map(element -> convertToTarget(element, fieldStructure))
-                    .collect(Collectors.toList());
+            checkKind(listEvent, fieldStructure.getName(), KindCase.LIST_VALUE);
+
+            List<Value> valuesList = listEvent.getListValue().getValuesList();
+            int size = valuesList.size();
+            List<Object> result = new ArrayList<>(size);
+            for (int i = 0; i < size; i++) {
+                Value element = valuesList.get(i);
+                try {
+                    result.add(convertToTarget(element, fieldStructure));
+                } catch (RuntimeException e) {
+                    throw new MessageConvertException("[" + i + "]", e);
+                }
+            }
+            return result;
```

Comment:
This whole method is very similar to the `convertComplexList` method. Maybe we need to extract the common logic to a separate method and pass the action on the value using lambda.

```java
private <T> List<T> convertList(ListValue listValue, IFieldStructure fieldStructure, BiFunction<Value, IFieldStructure, T> mapper)  {
    //...
    result.add(mapper.apply(element, fieldStructure));
}
```
or
```java
private <T> List<T> convertList(ListValue listValue, Function<Value, T> mapper)  {
    //...
    result.add(mapper.apply(element));
}
```